### PR TITLE
【Hackathon No58】-fix atan2

### PR DESCRIPTION
**Pull request by enkilee**
### PR types

Others
### PR changes

APIs
### Describe

fix atan2

### CODE REVIEW

File: paddle/phi/kernels/atan2_grad_kernel.h [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #pragma once
 
 #include "paddle/phi/core/dense_tensor.h"
+#include "paddle/phi/core/device_context.h"
```
Comment (ZzSean) [inline paddle/phi/kernels/atan2_grad_kernel.h]: 这里不用加吧

File: python/paddle/fluid/tests/unittests/test_atan2_op.py [hunk 2]
```diff
@@ -130,6 +130,23 @@ def run(place):
             run(place)
 
 
+class TestAtan2OpBF16(unittest.TestCase):
+    def testatan2bf16(OpTest):
```
Comment (ZzSean) [inline python/paddle/fluid/tests/unittests/test_atan2_op.py]: 这个写法有点问题，建议直接参考TestAtan2来写

Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_atan2_op.py]: 嗯，参考上面的写了，试试能不能过

File: paddle/phi/kernels/atan2_grad_kernel.h (hunk 1 above)
Comment (enkilee) [reply paddle/phi/kernels/atan2_grad_kernel.h]: OK

**Comment by enkilee** [comment]
Not equal to tolerance rtol=0.04, atol=0.1
2023-03-12 20:53:32 Operator atan2 error, Gradient Check On Place(cpu) variable X1 (shape: (15, 17), dtype: <class 'numpy.uint16'>) max gradient diff over limit
2023-03-12 20:53:32 Mismatched elements: 5 / 255 (1.96%)
2023-03-12 20:53:32 Max absolute difference: 0.82210501
2023-03-12 20:53:32 Max relative difference: 17.1498574
2023-03-12 20:53:32  x: array([[0.00351 , 0.002609, 0.002441, 0.001801, 0.005524, 0.003143,
2023-03-12 20:53:32         0.015869, 0.003418, 0.002914, 0.002808, 0.002029, 0.004852,
2023-03-12 20:53:32         0.002625, 0.001396, 0.003036, 0.005829, 0.0019  ],...
2023-03-12 20:53:32  y: array([[0.002743, 0.001307, 0.001096, 0.001641, 0.002259, 0.005355,
2023-03-12 20:53:32         0.001338, 0.001244, 0.003658, 0.153945, 0.005613, 0.005016,
2023-03-12 20:53:32         0.001333, 0.002405, 0.00104 , 0.001005, 0.003632],...

想问下，像这种 绝对误差阈值 超了的，该如何着手分析？

File: paddle/phi/kernels/cpu/atan2_grad_kernel.cc [hunk 3]
```diff
@@ -22,4 +22,5 @@ PD_REGISTER_KERNEL(atan2_grad,
                    phi::Atan2GradKernel,
                    float,
                    double,
-                   phi::dtype::float16) {}
+                   phi::dtype::float16,
+                   phi::dtype::bfloat16) {}
```
Comment (ZzSean) [inline paddle/phi/kernels/cpu/atan2_grad_kernel.cc]: 此次更改不要求改CPU

File: python/paddle/fluid/tests/unittests/test_atan2_op.py [hunk 4]
```diff
@@ -130,6 +130,41 @@ def run(place):
...
+    not core.is_compiled_with_cuda()
+    or not core.is_bfloat16_supported(core.CUDAPlace(0)),
+    "core is not compiled with CUDA and not support the bfloat16",
+)
+class TestAtan2OpBf16(OpTest):
+    def setUp(self):
+        self.op_type = 'atan2'
+        self.python_api = paddle.atan2
+        self.dtype = np.uint16
+        self.__class__.op_type = self.op_type
```
Comment (ZzSean) [inline python/paddle/fluid/tests/unittests/test_atan2_op.py]: 这句可以删，下面的test_check_output和test_check_grad中直接调用check_output_with_place ,只跑GPU的测试就可以

File: python/paddle/fluid/tests/unittests/test_atan2_op.py [hunk 5]
```diff
@@ -130,6 +130,41 @@ def run(place):
...
+        x2 = np.random.uniform(0.1, 1, [15, 17]).astype('float32')
+        out = np.arctan2(x1, x2)
+
+        self.inputs = {
+            'X1': convert_float_to_uint16(x1),
+            'X2': convert_float_to_uint16(x2),
+        }
+        self.outputs = {'Out': convert_float_to_uint16(out)}
+
+    def test_check_grad(self):
```
Comment (ZzSean) [inline python/paddle/fluid/tests/unittests/test_atan2_op.py:156]: 少了test_check_output

File: python/paddle/fluid/tests/unittests/test_atan2_op.py [hunk 6]
```diff
@@ -130,6 +130,41 @@ def run(place):
...
+        if self.dtype not in [np.int32, np.int64]:
+            self.check_grad(
+                ['X1', 'X2'],
+                'Out',
+                user_defined_grads=atan2_grad(
```
Comment (ZzSean) [inline python/paddle/fluid/tests/unittests/test_atan2_op.py]: 可以不用自定义user_defined_grads，尝试下能否通过

**Review by ZzSean — APPROVED** [review]
LGTM

**Pull request merged**